### PR TITLE
[VMware] Disconnect/Detach config drive ISO (if exists) on stop VM

### LINKED ISSUE #9453: ConfigDrive ISO is not detached from stopped on VMware
steps to reproduce the issue

- create L2 network with ConfigDrive

- create a vm on the network

ACS created a ConfigDrive ISO and attached to the VM on creation

```
2024-07-26T07:36:08,710 DEBUG [c.c.a.t.Request] (Work-Job-Executor-22:[ctx-901ad6df, job-185/job-186, ctx-f987f596]) (logid:0f2a1e0f) Seq 3-48413695994248261: Sending  { Cmd , MgmtId: 32989425697365, via: 3(s-2-VM), Ver: v1, Flags: 100011, [{"com.cloud.agent.api.HandleConfigDriveIsoCommand":{"isoFile":"configdrive/i-2-24-VM.iso","create":"true","destStore":{"com.cloud.agent.api.to.NfsTO":{"_url":"NFS://10.0.32.4/acs/secondary/pr9346-t10845-vmware-80/pr9346-t10845-vmware-80-sec1","_role":"Image"}},"useHostCacheOnUnsupportedPool":"true","preferHostCache":"false","wait":"0","bypassHostMaintenance":"false"}}] }
2024-07-26T07:36:09,256 DEBUG [c.c.a.m.ClusteredAgentManagerImpl] (Work-Job-Executor-22:[ctx-901ad6df, job-185/job-186, ctx-f987f596]) (logid:0f2a1e0f) Details from executing class com.cloud.agent.api.HandleConfigDriveIsoCommand: Successfully saved config drive at secondary storage
```

- stop the vm

ACS delete the ConfigDrive ISO

```
2024-07-26T08:06:35,901 DEBUG [c.c.a.t.Request] (Work-Job-Executor-23:[ctx-131bbbc2, job-187/job-188, ctx-d832a569]) (logid:451911a4) Seq 3-48413695994248293: Sending  { Cmd , MgmtId: 32989425697365, via: 3(s-2-VM), Ver: v1, Flags: 100011, [{"com.cloud.agent.api.HandleConfigDriveIsoCommand":{"isoFile":"configdrive/i-2-24-VM.iso","create":"false","destStore":{"com.cloud.agent.api.to.NfsTO":{"_url":"NFS://10.0.32.4/acs/secondary/pr9346-t10845-vmware-80/pr9346-t10845-vmware-80-sec1","_role":"Image"}},"useHostCacheOnUnsupportedPool":"false","preferHostCache":"false","wait":"0","bypassHostMaintenance":"false"}}] }
```

However, on vCenter the VM still has the ISO attached, which will causes some problem e.g. during VM snapshotting

image

##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~

~~~

##### CLOUDSTACK VERSION

~~~
4.19
~~~

##### OS / ENVIRONMENT

vmware 8.0b

##### STEPS TO REPRODUCE

~~~

~~~

##### EXPECTED RESULTS

~~~

~~~

##### ACTUAL RESULTS

~~~

~~~

### PR DESCRIPTION
**Pull request by sureshanaparti**
### Description

This PR disconnects/detachs config drive ISO (if exists) on stop VM in VMware. This is required for support some VM operations involving export OVF, on stopped VMs.

Root cause: Export OVF fails as there is no config drive file, it is removed when VM is stopped.

Fixes #9453

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI
- [ ] test (unit or integration test code)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [x] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?

Manually tested the start,stop,start VM operations and VM instance volume snapshots with config drive on VMware.

**Before =>**

Before_ConfigDrive_NotDetached_VMware

Before_ConfigDrive_NotDetached_And_Snapshot_CS

**After =>**

After_ConfigDrive_Detached_VMware

After_ConfigDrive_Detached_And_Snapshot_CS

```
2024-07-29T15:56:17,721 DEBUG [c.c.a.t.Request] (Work-Job-Executor-2:[ctx-bd90a732, job-217/job-218, ctx-bd64bbee]) (logid:fd80b170) Seq 1-234468655599976463: Executing:  { Cmd , MgmtId: 32989425697365, via: 1(10.0.34.85), Ver: v1, Flags: 100011, [{"com.cloud.agent.api.StopCommand":{"isProxy":"false","checkBeforeCleanup":"false","forceStop":"false","vlanToPersistenceMap":{"1727":"true"},"volumesToDisconnect":[],"vmName":"i-2-25-VM","executeInSequence":"false","wait":"0","bypassHostMaintenance":"false"}}] }
...
2024-07-29T15:56:21,045 INFO  [c.c.h.v.r.VmwareResource] (DirectAgent-14:[ctx-7ac38317, 10.0.34.85, job-217/job-218, cmd: StopCommand]) (logid:fd80b170) Disconnecting config drive at location: [c25a36f2689731989d17a452538ae70d] configdrive/i-2-25-VM.iso

2024-07-29T15:56:21,253 DEBUG [c.c.n.e.ConfigDriveNetworkElement] (Work-Job-Executor-2:[ctx-bd90a732, job-217/job-218, ctx-bd64bbee]) (logid:fd80b170) Deleting config drive ISO for vm: i-2-25-VM
2024-07-29T15:56:21,256 DEBUG [c.c.a.m.ClusteredAgentAttache] (Work-Job-Executor-2:[ctx-bd90a732, job-217/job-218, ctx-bd64bbee]) (logid:fd80b170) Seq 2-4190599453268246540: Routed from 32989425697365
2024-07-29T15:56:21,257 DEBUG [c.c.a.t.Request] (Work-Job-Executor-2:[ctx-bd90a732, job-217/job-218, ctx-bd64bbee]) (logid:fd80b170) Seq 3-4190599453268246540: Sending  { Cmd , MgmtId: 32989425697365, via: 3(s-2-VM), Ver: v1, Flags: 100011, [{"com.cloud.agent.api.HandleConfigDriveIsoCommand":{"isoFile":"configdrive/i-2-25-VM.iso","create":"false","destStore":{"com.cloud.agent.api.to.NfsTO":{"_url":"NFS://10.0.32.4/acs/secondary/pr9346-t10845-vmware-80/pr9346-t10845-vmware-80-sec1","_role":"Image"}},"useHostCacheOnUnsupportedPool":"false","preferHostCache":"false","wait":"0","bypassHostMaintenance":"false"}}] }
2024-07-29T15:56:21,321 DEBUG [c.c.a.t.Request] (AgentManager-Handler-8:[]) (logid:) Seq 3-4190599453268246540: Processing:  { Ans: , MgmtId: 32989425697365, via: 3, Ver: v1, Flags: 10, [{"com.cloud.agent.api.HandleConfigDriveIsoAnswer":{"result":"true","wait":"0","bypassHostMaintenance":"false"}}] }
2024-07-29T15:56:21,321 DEBUG [c.c.a.t.Request] (Work-Job-Executor-2:[ctx-bd90a732, job-217/job-218, ctx-bd64bbee]) (logid:fd80b170) Seq 3-4190599453268246540: Received:  { Ans: , MgmtId: 32989425697365, via: 3(s-2-VM), Ver: v1, Flags: 10, { HandleConfigDriveIsoAnswer } }
...
2024-07-29T15:56:21,213 DEBUG [c.c.a.t.Request] (DirectAgent-14:[ctx-7ac38317]) (logid:fd80b170) Seq 1-234468655599976463: Processing:  { Ans: , MgmtId: 32989425697365, via: 1(10.0.34.85), Ver: v1, Flags: 10, [{"com.cloud.agent.api.StopAnswer":{"result":"true","details":"Stop VM i-2-25-VM Succeed","wait":"0","bypassHostMaintenance":"false"}}] }
2024-07-29T15:56:21,213 DEBUG [c.c.a.t.Request] (Work-Job-Executor-2:[ctx-bd90a732, job-217/job-218, ctx-bd64bbee]) (logid:fd80b170) Seq 1-234468655599976463: Received:  { Ans: , MgmtId: 32989425697365, via: 1(10.0.34.85), Ver: v1, Flags: 10, { StopAnswer } }
```

### CODE REVIEW

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Review by weizhouapache — APPROVED** [review]
code lgtm

not tested yet

**Comment by weizhouapache** [comment]
@blueorangutan test rocky8 vmware-70u3

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test ol8 vmware-70u3

**Comment by sureshanaparti** [comment]
@blueorangutan package

File: vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/VirtualMachineMO.java [hunk 1]
```diff
@@ -3203,6 +3203,19 @@ public VirtualDevice getIsoDevice() throws Exception {
...
+        List<VirtualDevice> isoDevices = new ArrayList<>();
+        List<VirtualDevice> devices = _context.getVimClient().getDynamicProperty(_mor, "config.hardware.device");
+        if (devices != null && devices.size() > 0) {
+            for (VirtualDevice device : devices) {
+                if (device instanceof VirtualCdrom) {
+                    isoDevices.add(device);
+                }
+            }
+        }
+        return isoDevices;
```
Comment (shwstppr) [inline vmware-base/src/main/java/com/cloud/hypervisor/vmware/mo/VirtualMachineMO.java:3216]:
```suggestion
        List<VirtualDevice> devices = _context.getVimClient().getDynamicProperty(_mor, "config.hardware.device");
        if (CollectionUtils.isEmpty(devices)) {
            return new ArrayList<>();
        }
        return devices.stream().filter(d -> d instanceof VirtualCdrom).collect(Collectors.toList());
```

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java [hunk 2]
```diff
@@ -4466,6 +4470,28 @@ protected Answer execute(StopCommand cmd) {
...
+
+            for (VirtualDevice isoDevice : isoDevices) {
+                if(isoDevice.getBacking() instanceof VirtualCdromIsoBackingInfo) {
+                    String isoFilePath = ((VirtualCdromIsoBackingInfo)isoDevice.getBacking()).getFileName();
+                    if (isoFilePath.contains(ConfigDrive.CONFIGDRIVEDIR)) {
+                        s_logger.info(String.format("Disconnecting config drive at location: %s", isoFilePath));
+                        vmMo.detachIso(isoFilePath, true);
+                        return;
+                    }
+                }
```
Comment (shwstppr) [inline plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java:4488]:
```suggestion
                if(!(isoDevice.getBacking() instanceof VirtualCdromIsoBackingInfo)) {
                    continue;
                }
                String isoFilePath = ((VirtualCdromIsoBackingInfo)isoDevice.getBacking()).getFileName();
                if (!isoFilePath.contains(ConfigDrive.CONFIGDRIVEDIR)) {
                    continue;
                }
                s_logger.info(String.format("Disconnecting config drive at location: %s", isoFilePath));
                vmMo.detachIso(isoFilePath, true);
                return;
```

**Review by shwstppr — APPROVED** [review]
Logic looks good. Minor comments wrt refactoring

**Review by vladimirpetrov — APPROVED** [review]
LGTM based on manual testing using VMWare 8.0u1.

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test ol8 vmware-70u3

**Pull request merged**